### PR TITLE
Move some circleci tests to be Go tests

### PR DESCRIPTION
**Pull request by hugelgupf**
Moves all ./u-root tests in circleci to Go.

Just go to /u-root/ and type go test to try it. Takes ~3 mins to run on my machine.

### CODE REVIEW

File: pkg/cpio/archive.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+	r.Name = Normalize(r.Name)
+	if s, ok := a.Files[r.Name]; ok {
+		return Equal(r, s)
+	}
+	return false
+}
+
+// Get returns a record for the normalized path or false if there is none.
+//
+// The path is normalized using Normalize, so Get("/bin/bar") is the same as
```
Comment (GanShun) [inline pkg/cpio/archive.go]: what's Get("/bin/bar") the same as?

Comment (hugelgupf) [reply pkg/cpio/archive.go]: see.. the line below continued?

File: .circleci/config.yml [hunk 2]
```diff
@@ -131,10 +89,10 @@ jobs:
       - checkout
       - run:
           name: Test all
-          command: go test -a -ldflags '-s' ./pkg/... ./cmds/... ./xcmds/...
+          command: go test -a -ldflags '-s' ./pkg/... ./cmds/... ./xcmds/... .
```
Comment (rjoleary) [inline .circleci/config.yml]: Could we instead use this?

```suggestion
          command: go test -a -ldflags '-s' ./...
```

File: uroot_test.go [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+		fmt.Sprintf("GOROOT=%s", goroot))
+	out, err := c.CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("could not build go source %v; output\n%s", err, out)
+	}
+	return nil
+}
+
+func TestUrootCmdline(t *testing.T) {
+	t.Parallel()
```
Comment (rjoleary) [inline uroot_test.go]: This doesn't make the TestUrootCmdline test any faster. See https://golang.org/pkg/testing/#hdr-Subtests_and_Sub_benchmarks

File: u-root.go [hunk 4]
```diff
@@ -335,21 +108,72 @@ func Main() error {
...
+		}
 	}
 
-	opts := uroot.Opts{
-		Env: env,
+	var (
+		c           []uroot.Commands
+		initCommand = *initCmd
+	)
+	if !*noCommands {
```
Comment (rjoleary) [inline u-root.go]: What was the reason behind adding this flag? I doesn't seem needed by the rest of the PR.

File: uroot_test.go [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+			for _, v := range tt.validators {
+				if err := v.Validate(a); err != nil {
+					t.Errorf("validator failed: %v / archive:\n%s", err, a)
+				}
+			}
+
+			f2, sum2 := buildIt(t, tt.args, tt.env, tt.err)
+			defer os.RemoveAll(f2.Name())
+			if !bytes.Equal(sum1, sum2) {
+				t.Fatalf("not reproducible, hashes don't match")
```
Comment (rjoleary) [inline uroot_test.go]: Can you also print the path to the two temp files and not delete them on failure?

File: .circleci/config.yml (hunk 2 above)
Comment (hugelgupf) [reply .circleci/config.yml]: that'll run all tests in the vendor directory

File: u-root.go (hunk 4 above)
Comment (hugelgupf) [reply u-root.go]: to be able to test the -extra-files commands without having to compile anything. speeds up the tests.

**Pull request merged**